src/main/java/uk/ac/ebi/pride/spectracluster/engine/ReassessingIncrementalClusteringEngine.java

```diff
@@ -0,0 +1,432 @@
+     * these are dropped and will be handled as never modifies this pass
+     *
+     * @param precursorMz new Mz
+     * @return !null list of clusters to remove
+     */
+    protected List<ICluster> findClustersTooLow(double precursorMz) {
+        setCurrentMZ(precursorMz); // also performs sanity check whether precursorMz is larger than currentMz
+
+        double windowSize1 = getWindowSize();
+        double lowestMZ = precursorMz - windowSize1;
+        List<ICluster> clustersToremove = new ArrayList<ICluster>();
+        List<ICluster> myClusters = internalGetClusters();
+        for (ICluster test : myClusters) {
+            float testPrecursorMz = test.getPrecursorMz();
+            if (lowestMZ > testPrecursorMz) {
+                clustersToremove.add(test);
+            }
+        }
+        if (!clustersToremove.isEmpty())
+            internalGetClusters().removeAll(clustersToremove);   // might break hear
+
+        return clustersToremove;
+
+    }
+
+
+    /**
+     * this method is called by guaranteeClean to place any added clusters in play
+     * for further clustering
+     */
+    protected void addToClusters(final ICluster clusterToAdd) {
+        List<ICluster> myClusters = internalGetClusters();
+        if (myClusters.isEmpty()) {   // no checks just add
+            myClusters.add(new SpectralCluster(clusterToAdd, Defaults.getDefaultConsensusSpectrumBuilder()));
+            numberNotMerge++;
+            return;
+        }
+
+        //  handle the case of subsets and almost subsets
+        if (handleFullContainment(clusterToAdd))
+            return; // no need to add we are contained
+
+
+        ISimilarityChecker theSimilarityChecker = getSimilarityChecker();
+        List<ISpectrum> clusteredSpectraToAdd = clusterToAdd.getClusteredSpectra();
+
+        double highestSimilarityScore = 0;
+        ICluster mostSimilarCluster = null;
+
+        ISpectrum consensusSpectrum1 = clusterToAdd.getConsensusSpectrum();  // subspectra are really only one spectrum clusters
+        // find the cluster with the highest similarity score
+        for (ICluster cluster : myClusters) {
+            ISpectrum consensusSpectrum = cluster.getConsensusSpectrum();
+
+            double similarityScore = theSimilarityChecker.assessSimilarity(consensusSpectrum, consensusSpectrum1);
+
+            if (similarityScore > highestSimilarityScore && similarityScore >= similarityThreshold) {
+                highestSimilarityScore = similarityScore;
+                mostSimilarCluster = cluster;
+            }
+        }
+
+        if (mostSimilarCluster != null) {
+            // add to cluster
+            ISpectrum[] clusteredSpectra = new ISpectrum[clusteredSpectraToAdd.size()];
+            final ISpectrum[] merged = clusteredSpectraToAdd.toArray(clusteredSpectra);
+            mostSimilarCluster.addSpectra(merged);
+            mostSimilarCluster.setProperty(CLUSTER_PROPERTY_IS_DIRTY, "");
```

Comment:
Good idea on setting the cluster as dirty, however, I am not sure using empty string is the best practice here. If it is a boolean, I would recommend to use boolean.